File: android/app/src/main/java/com/celalbasaran/stripmate/data/model/Streak.kt
package com.celalbasaran.stripmate.data.model

import com.google.firebase.Timestamp
import com.google.firebase.firestore.DocumentSnapshot
import java.util.Calendar
import java.util.Date

data class Streak(
    val id: String = "",
    val userIds: List<String> = emptyList(),
    val currentStreak: Int = 0,
    val longestStreak: Int = 0,
    val totalExchanges: Int = 0,
    val lastExchangeDate: Date = Date(),
    val lastSenderId: String = "",
    val friendshipScore: Int = 0
) {
    val isExpiringSoon: Boolean
        get() {
            if (currentStreak <= 0) return false
            val calendar = Calendar.getInstance()
            val today = calendar.apply {
                set(Calendar.HOUR_OF_DAY, 0)
                set(Calendar.MINUTE, 0)
                set(Calendar.SECOND, 0)
                set(Calendar.MILLISECOND, 0)
            }.timeInMillis

            calendar.time = lastExchangeDate
            val lastDay = calendar.apply {
                set(Calendar.HOUR_OF_DAY, 0)
                set(Calendar.MINUTE, 0)
                set(Calendar.SECOND, 0)
                set(Calendar.MILLISECOND, 0)
            }.timeInMillis

            val daysSince = ((today - lastDay) / (1000 * 60 * 60 * 24)).toInt()
            return daysSince >= 1
        }

    val friendshipTier: FriendshipTier
        get() = when {
            friendshipScore < 50 -> FriendshipTier.TANIDIK
            friendshipScore < 150 -> FriendshipTier.MUHABBET
            friendshipScore < 350 -> FriendshipTier.YAKIN
            friendshipScore < 700 -> FriendshipTier.SIRDAS
            else -> FriendshipTier.KADIM
        }

    val nextTierThreshold: Int
        get() = when (friendshipTier) {
            FriendshipTier.TANIDIK -> 50
            FriendshipTier.MUHABBET -> 150
            FriendshipTier.YAKIN -> 350
            FriendshipTier.SIRDAS -> 700
            FriendshipTier.KADIM -> 1000
        }

    val tierProgress: Double
        get() {
            val current = friendshipScore.toDouble()
            val thresholds = listOf(0 to 50, 50 to 150, 150 to 350, 350 to 700, 700 to 1000)
            for ((low, high) in thresholds) {
                if (friendshipScore < high) {
                    return (current - low) / (high - low)
                }
            }
            return 1.0
        }

    fun toMap(): Map<String, Any?> = buildMap {
        put("id", id)
        put("userIds", userIds)
        put("currentStreak", currentStreak)
        put("longestStreak", longestStreak)
        put("totalExchanges", totalExchanges)
        put("lastExchangeDate", Timestamp(lastExchangeDate))
        put("lastSenderId", lastSenderId)
        put("friendshipScore", friendshipScore)
    }

    companion object {
        fun streakId(uid1: String, uid2: String): String {
            return listOf(uid1, uid2).sorted().joinToString("_")
        }

        @Suppress("UNCHECKED_CAST")
        fun fromDocument(doc: DocumentSnapshot): Streak? {
            if (!doc.exists()) return null
            return Streak(
                id = doc.id,
                userIds = doc.get("userIds") as? List<String> ?: emptyList(),
                currentStreak = doc.getLong("currentStreak")?.toInt() ?: 0,
                longestStreak = doc.getLong("longestStreak")?.toInt() ?: 0,
                totalExchanges = doc.getLong("totalExchanges")?.toInt() ?: 0,
                lastExchangeDate = doc.getTimestamp("lastExchangeDate")?.toDate() ?: Date(),
                lastSenderId = doc.getString("lastSenderId") ?: "",
                friendshipScore = doc.getLong("friendshipScore")?.toInt() ?: 0
            )
        }
    }
}

enum class FriendshipTier(val tierName: String, val tierIcon: String) {
    TANIDIK("Tanıdık", "circle_dotted"),
    MUHABBET("Muhabbet", "cup_and_saucer"),
    YAKIN("Yakın", "link"),
    SIRDAS("Sırdaş", "key"),
    KADIM("Kadim", "infinity")
}

File: StripMate/Models/Streak.swift
import Foundation

/// Represents the streak (consecutive daily exchange) between two users.
/// Stored in Firestore at: `streaks/{streakId}` where streakId = sorted "uid1_uid2"
public struct Streak: Identifiable, Codable, Sendable {
    /// Composite id: sorted "uid1_uid2"
    public let id: String
    /// The two user IDs participating in this streak
    public let userIds: [String]
    /// Current consecutive-day count
    public var currentStreak: Int
    /// All-time longest streak between these two users
    public var longestStreak: Int
    /// Total strips exchanged (both directions combined)
    public var totalExchanges: Int
    /// Last date a strip was exchanged (used to check continuity)
    public var lastExchangeDate: Date
    /// The userId who sent the last strip (to show "your turn" indicator)
    public var lastSenderId: String
    /// Friendship score: weighted combination of streak, frequency, recency
    public var friendshipScore: Int
    /// Whether either side has used their weekly freeze on the current cycle.
    /// Resets every Monday (server-side via weeklyFreezeReset cron).
    public var freezeUsedThisWeek: Bool = false
    /// Optional timestamp until which the streak is "frozen" — backend treats
    /// these days as a continuation rather than a break. Cleared when both
    /// users exchange again.
    public var frozenUntil: Date? = nil

    /// Whether the streak is about to expire (no exchange today yet, streak > 0)
    public var isExpiringSoon: Bool {
        guard currentStreak > 0 else { return false }
        // A frozen streak is in safe-mode — don't surface the expiring badge.
        if let frozenUntil, frozenUntil > Date() { return false }
        let calendar = Calendar.current
        let lastDay = calendar.startOfDay(for: lastExchangeDate)
        let today = calendar.startOfDay(for: Date())
        let daysSince = calendar.dateComponents([.day], from: lastDay, to: today).day ?? 0
        return daysSince >= 1 // Haven't exchanged today
    }

    /// True when a freeze is currently active (frozenUntil is in the future).
    public var isFrozen: Bool {
        guard let frozenUntil else { return false }
        return frozenUntil > Date()
    }

    /// True when the user can still freeze this week and the streak is at risk.
    public var canFreezeNow: Bool {
        currentStreak > 0
            && !freezeUsedThisWeek
            && !isFrozen
            && isExpiringSoon
    }

    /// Helper to build the composite streak id from two user ids
    public static func streakId(for uid1: String, and uid2: String) -> String {
        [uid1, uid2].sorted().joined(separator: "_")
    }

    public init(
        id: String,
        userIds: [String],
        currentStreak: Int = 0,
        longestStreak: Int = 0,
        totalExchanges: Int = 0,
        lastExchangeDate: Date = Date(),
        lastSenderId: String = "",
        friendshipScore: Int = 0,
        freezeUsedThisWeek: Bool = false,
        frozenUntil: Date? = nil
    ) {
        self.id = id
        self.userIds = userIds
        self.currentStreak = currentStreak
        self.longestStreak = longestStreak
        self.totalExchanges = totalExchanges
        self.lastExchangeDate = lastExchangeDate
        self.lastSenderId = lastSenderId
        self.friendshipScore = friendshipScore
        self.freezeUsedThisWeek = freezeUsedThisWeek
        self.frozenUntil = frozenUntil
    }
}

// MARK: - Friendship Score Tiers

public extension Streak {
    /// Tier based on friendship score
    enum FriendshipTier: String, Sendable {
        case tanidik = "Tanıdık"
        case muhabbet = "Muhabbet"
        case yakin = "Yakın"
        case sirdas = "Sırdaş"
        case kadim = "Kadim"

        /// Türkçe seviye adı
        public var tierName: String { rawValue }

        /// Monokrom SF Symbol ikon adı
        public var tierIcon: String {
            switch self {
            case .tanidik:  return "circle.dotted"
            case .muhabbet: return "cup.and.saucer.fill"
            case .yakin:    return "link"
            case .sirdas:   return "key.fill"
            case .kadim:    return "infinity"
            }
        }
    }

    var tier: FriendshipTier {
        switch friendshipScore {
        case 0..<50:   return .tanidik
        case 50..<150: return .muhabbet
        case 150..<350: return .yakin
        case 350..<700: return .sirdas
        default:        return .kadim
        }
    }

    /// Next tier threshold (for progress bar)
    var nextTierThreshold: Int {
        switch tier {
        case .tanidik:  return 50
        case .muhabbet: return 150
        case .yakin:    return 350
        case .sirdas:   return 700
        case .kadim:    return 1000
        }
    }

    /// Progress towards the next tier (0.0 - 1.0)
    var tierProgress: Double {
        let current = Double(friendshipScore)
        let thresholds: [(Int, Int)] = [(0, 50), (50, 150), (150, 350), (350, 700), (700, 1000)]
        for (low, high) in thresholds {
            if friendshipScore < high {
                return (current - Double(low)) / Double(high - low)
            }
        }
        return 1.0
    }
}
